controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -310,8 +319,60 @@ func (r *StorageClassClaimReconciler) reconcileConsumerPhases() (reconcile.Resul
 		// Update the StorageClassClaim status.
 		r.storageClassClaim.Status.Phase = v1alpha1.StorageClassClaimReady
 
+	} else if r.storageClassClaim.UID != "" {
+
+		// Deletion phase.
+		// Update the StorageClassClaim status.
+		r.storageClassClaim.Status.Phase = v1alpha1.StorageClassClaimDeleting
+
+		// Call `RevokeStorageClassClaim` service on the provider server with StorageClassClaim as a request message.
+		// Check if StorageClassClaim is still exists (it might have been manually removed during the StorageClass
+		// removal above).
+		_, err = providerClient.RevokeStorageClassClaim(r.ctx, r.storageCluster.Status.ExternalStorage.ConsumerID, r.storageClassClaim.Name)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// Delete StorageClass.
+		// Make sure there are no StorageClass consumers left.
+		// Check if StorageClass is in use, if yes, then fail.
+		// Wait until all PVs using the StorageClass under deletion are removed.
+		// Check for any PVs using the StorageClass.
+		pvList := corev1.PersistentVolumeList{}
+		err := r.list(&pvList, client.MatchingFields{
+			"spec.storageClassName": r.storageClassClaim.Name,
+		})
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to list PersistentVolumes: %s", err)
+		}
+		if len(pvList.Items) > 0 {
+			return reconcile.Result{}, fmt.Errorf("StorageClass %s is still in use by one or more PV(s)",
+				r.storageClassClaim.Name)
+		}
+		storageClass := &storagev1.StorageClass{}
+		storageClass.Name = r.storageClassClaim.Name
+		if err = r.get(storageClass); err == nil {
+
+			// Remove finalizer from StorageClass.
+			storageClass.SetFinalizers(remove(storageClass.GetFinalizers(), StorageClassFinalizer))
+			if err := r.update(storageClass); err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed to remove finalizer from StorageClass: %s", err)
+			}
+			if err = r.delete(storageClass); err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed to delete StorageClass %s: %s", storageClass.Name, err)
```

Comment:
We set a finalizer on the storage class, but we do not react to the deletion of storage classes. 
To me it does not make sense, I think the behavior we want is to recreate the storage class in case it got deleted (by a 3rd agent) and we cannot do so if we have a finalizer as the storage class will not get deleted and we will not reinvoke the creation path. 

I think we need to get rid of this finalizer and at the same time validate our creation path is handling recreation in case of an event that was raised as a result of storage class deletion. My guess it is, but we do need to validate.